demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1524 @@
+    MQTTStatus_t mqttStatus;
+    MQTTContext_t * pMqttContext = &mqttContext;
+    MQTTSubscribeInfo_t pSubscriptionList[ 1 ];
+
+    assert( pMqttContext != NULL );
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength > 0 );
+
+    /* Start with everything at 0. */
+    ( void ) memset( ( void * ) pSubscriptionList, 0x00, sizeof( pSubscriptionList ) );
+
+    /* Set the QoS , topic and topic length. */
+    pSubscriptionList[ 0 ].qos = qos;
+    pSubscriptionList[ 0 ].pTopicFilter = pTopicFilter;
+    pSubscriptionList[ 0 ].topicFilterLength = topicFilterLength;
+
+    if( pthread_mutex_lock( &mqttMutex ) == 0 )
+    {
+        /* Send SUBSCRIBE packet. */
+        mqttStatus = MQTT_Subscribe( pMqttContext,
+                                     pSubscriptionList,
+                                     sizeof( pSubscriptionList ) / sizeof( MQTTSubscribeInfo_t ),
+                                     MQTT_GetPacketId( pMqttContext ) );
+
+        pthread_mutex_unlock( &mqttMutex );
+    }
+    else
+    {
+        LogError( ( "Failed to acquire mqtt mutex for executing MQTT_Subscribe"
+                    ",errno=%s",
+                    strerror( errno ) ) );
+    }
+
+    if( mqttStatus != MQTTSuccess )
+    {
+        LogError( ( "Failed to send SUBSCRIBE packet to broker with error = %u.",
+                    mqttStatus ) );
+
+        otaRet = OtaMqttSubscribeFailed;
+    }
+    else
+    {
+        LogInfo( ( "SUBSCRIBE topic %.*s to broker.\n\n",
+                   topicFilterLength,
+                   pTopicFilter ) );
+    }
+
+    /* Register callback to subscription manager. */
+    subscriptionStatus = SubscriptionManager_RegisterCallback( pTopicFilter, topicFilterLength, callback );
+
+    if( subscriptionStatus != SUBSCRIPTION_MANAGER_SUCCESS )
+    {
+        LogWarn( ( "Failed to register a callback to subscription manager with error = %d.",
+                   subscriptionStatus ) );
+    }
+
+    return otaRet;
+}
+
+/*-----------------------------------------------------------*/
+
+static OtaMqttStatus_t mqttPublish( const char * const pacTopic,
+                                    uint16_t topicLen,
+                                    const char * pMsg,
+                                    uint32_t msgSize,
+                                    uint8_t qos )
+{
+    OtaMqttStatus_t otaRet = OtaMqttSuccess;
+
+    MQTTStatus_t mqttStatus = MQTTBadParameter;
+    MQTTPublishInfo_t publishInfo;
+    MQTTContext_t * pMqttContext = &mqttContext;
+
+    /* Set the required publish parameters. */
+    publishInfo.pTopicName = pacTopic;
+    publishInfo.topicNameLength = topicLen;
+    publishInfo.qos = qos;
+    publishInfo.pPayload = pMsg;
+    publishInfo.payloadLength = msgSize;
+
+    if( pthread_mutex_lock( &mqttMutex ) == 0 )
+    {
+        mqttStatus = MQTT_Publish( pMqttContext,
+                                   &publishInfo,
+                                   MQTT_GetPacketId( pMqttContext ) );
```

Comment:
Do we ever use QoS 0?
```suggestion
                                   ( qos ) ? MQTT_GetPacketId( pMqttContext ) : 0U );
```